lib/Controller/SubsonicController.php

```diff
@@ -1067,17 +1071,64 @@ protected function deleteBookmark(string $id) : array {
 	/**
 	 * @SubsonicAPI
 	 */
-	protected function getPlayQueue() : array {
-		// TODO: not supported yet
-		return ['playQueue' => []];
+	protected function getPlayQueue() : Response {
+		/** @var array|false $playQueue */
+		$playQueue = json_decode($this->configManager->getUserValue($this->user(), $this->appName, 'play_queue', 'false'), true);
+
+		if (!$playQueue) {
+			return $this->subsonicResponse([]);
+		}
+
+		$parsedEntries = \array_map([self::class, 'parseEntityId'], $playQueue['entry']);
+		$trackEntries = \array_filter($parsedEntries,fn ($parsedEntry) => $parsedEntry[0] === 'track');
+		$tracksById = ArrayUtil::createIdLookupTable($this->trackBusinessLayer->findById(
+			\array_map(fn ($trackEntry) => $trackEntry[1], $trackEntries), $this->user()
+		));
+		$apiTracks = $this->tracksToApi($tracksById);
+
+		$playQueue['entry'] = \array_reduce(
+			$parsedEntries,
+			function ($pqEntries, $parsedEntry) use ($apiTracks) {
+				[$type, $id] = $parsedEntry;
+				if ($type === 'track' && isset($apiTracks[$id])) {
+					$pqEntries[] = $apiTracks[$id];
+				} else if ($type === 'podcast_episode') {
+					try {
+						$pqEntries[] = $this->podcastEpisodeBusinessLayer->find($id, $this->user())->toSubsonicApi();
+					} catch (\Throwable $t) {
+						// catch missing podcast episode exceptions; maybe episode was deleted after being added to queue
+					}
+				}
+				return $pqEntries;
+			},
+			[]
+		);
+
+		return $this->subsonicResponse(['playQueue' => $playQueue]);
 	}
 
 	/**
 	 * @SubsonicAPI
 	 */
-	protected function savePlayQueue() : array {
-		// TODO: not supported yet
-		return [];
+	protected function savePlayQueue(array $id, string $c, string $u, ?string $current = null, ?int $position = null) : Response {
+		$changedDateTime = new \DateTime();
+		$playQueue = array_filter([
+			'entry' => array_filter(
+				$id,
+				fn (string $entityId) => in_array(self::parseEntityId($entityId)[0], ['track', 'podcast_episode'])
+			),
+			'changedBy' => $c,
+			'position' => $position,
+			'current' => $current,
+			/** @see Util::formatZuluDateTime (if only we could pass a datetime!) */
+			'changed' => $changedDateTime->format('Y-m-d\TH:i:s.v\Z'),
+			'username' => $u
```

Comment:
We should use `$this->user()` here instead of the URL argument `$u`. The reason is that the Music app supports the OpenSubsonic extension [API Key Authentication](https://opensubsonic.netlify.app/docs/extensions/apikeyauth/) which means that the client may pass the argument `apiKey` instead of `u` and `p`. The function will return the correct username in either case.

Another small difference is that the URL argument `$u` is case insensitive but `$this->user()` always contains the "proper" casing i.e. the same format in which the username was given when the account was created.